### PR TITLE
Issue #3902: Formatting of code chunks in html documentation (linewrap or scrolling)

### PR DESCRIPTION
**Pull request by sagar-shah94**
Issue: #3902

### CODE REVIEW

**Comment by sagar-shah94** [comment]
@rnveach @romani fixed issue with css

**Comment by rnveach** [comment]
@sagar-shah94 Please keep all work in 1 PR so we can see history. Previous PR is https://github.com/checkstyle/checkstyle/pull/3912 .
Please fix failing CIs.

**Comment by sagar-shah94** [comment]
@romani @rnveach 
please verify this pull request.

File: src/site/resources/css/site.css [hunk 1]
```diff
@@ -49,3 +49,10 @@ h2:hover .anchor {
 .anchor {
   display: none;
 }
+
+.source pre {
+ white-space: pre-wrap !important;
```
Comment (rnveach) [inline src/site/resources/css/site.css]: Can we ident these lines the same as the others. 2 spaces instead of 1.

**Comment by rnveach** [comment]
@romani I test on Firefox and Internet Explorer 11 and the change works fine for me.
Once minor issue is fixed, I am fine to merge.

**Comment by sagar-shah94** [comment]
@romani @rnveach when i ran mvn clean verify it is showing **BUILD SUCCESS** but here wercker/build is failing .Is it error or what?

**Comment by Vladlis** [comment]
@sagar-shah94 
no exception test on guava is timed out during `git clone`, please try triggering it again

**Comment by sagar-shah94** [comment]
How can i solve ?should i clone checkstyle repository again that is what you are suggesting?
@Vladlis

**Comment by rnveach** [comment]
@sagar-shah94 amend your commit with no changes and force push to branch.
Github will see this as a change even though nothing changed and force CIs to re-run.

We cannot manually re-run wercker as I just tried. It auto fails with a bogus message. It is a known issue with wercker.

**Comment by sagar-shah94** [comment]
OK thanks for help :)

**Comment by sagar-shah94** [comment]
All checks passed
@rnveach @romani

**Comment by romani** [comment]
in general concept is good, but in most cases we can not do linewrap as we rely of exact and strict formatting of code.
screenshot-non-wrapped
screenshot-wrapped

It is ok to wrap command lines , and some examples that are not sensitive to formatting.
@sagar-shah94 , can we apply your CCS to only certain `source` blocks ? we need to put some attribute on them to distinguish.

**Comment by rnveach** [comment]
@romani in ones we shouldn't wrap, we can force them to have a scrollbar instead of going off the page like originally

**Comment by romani** [comment]
yes , good idea.

**Comment by sagar-shah94** [comment]
@rnveach @romani 
what kind of changes do i have to make?

**Comment by rnveach** [comment]
@sagar-shah94 
First we have to find out if it is possible to mark `source`s with a `class` attribute so we can apply different CSS rules to different sources.
You can test this out locally. 

Next we need to create 2 different CSS rules for different `source`s:
1) should be wrapped to the next line if they try to go off the page
2) should have a scrollbar if the content tries to go off the page.
We already have type 1 in this PR. So we just need to create type 2.

Finally, we need to identify for every source in our XDOC which type of CSS should be applied to it.
@romani 's example of an Tree output is type 2.
Issue's example of `[ERROR:0] Javadoc comment at column 1 has...` should be type 1 because it is similar to a paragraph and is ok to wrap imo.

@romani you agree?
I don't know if we can write a test to automatically identify the type of wrapping we want for any `source`.
It depends on if we identify any non-tree sources as type 2.

**Comment by romani** [comment]
> romani you agree?

yes, all is correct.

> I don't know if we can write a test to automatically identify the type of wrapping we want for any source.

I doubt. Most case will be "type 2" in Checks descriptions.
command lie examples will be "type 1" - http://checkstyle.sourceforge.net/cmdline.html
It is human criteria, if content of `source` is harder to read/understand on linewrap - we should not do this.

**Comment by sagar-shah94** [comment]
Should i work on this issue?
I think i found solution we could use iframe tag instead of source tag to put scrollbar in bottom.
what do you suggest?
@rnveach @romani

**Comment by rnveach** [comment]
> Should i work on this issue?

Please do.

> I think i found solution we could use iframe tag instead of source tag to put scrollbar in bottom.

I am not against it, but is there any reason we can't use pure CSS?

**Comment by sagar-shah94** [comment]
Sir we can use pure css ,let me try with pure css and i will push code and let you know.

**Comment by sagar-shah94** [comment]
@rnveach @romani fixed css issue.scrollbar will come in source.it will not be wrapped.

**Comment by rnveach** [comment]
@sagar-shah94 How come there is only 1 CSS in code?

https://github.com/checkstyle/checkstyle/pull/3998#issuecomment-287574694
> Next we need to create 2 different CSS rules for different sources:
> 1. should be wrapped to the next line if they try to go off the page
> 2. should have a scrollbar if the content tries to go off the page.
> We already have type 1 in this PR. So we just need to create type 2.

We said earlier we would have 2, what you originally had in this PR plus this new one you just added.

**Comment by sagar-shah94** [comment]
but if we try to wrap content we have to apply css on pre tag and after applying that css scrollbar css was not working because that is nowrap css which was to be applied to parent div tag
@rnveach

**Comment by rnveach** [comment]
One last question then.

> but if we try to wrap content we have to apply css on pre tag and after applying that css scrollbar css was not working 


What about the first step I mentioned, were we don't apply both CSS rules to the same `source` tag, which I assume is what you are saying it happening. We separate them by `class` attribute.

> First we have to find out if it is possible to mark sources with a class attribute so we can apply different CSS rules to different sources.

**Comment by sagar-shah94** [comment]
if we want content to wrap we have to apply css on pre tag which is child tag of source .
if we want scrollbar horizontally we have to apply css on source tag which is parent tag.

@rnveach

**Comment by sagar-shah94** [comment]
image
Even github uses same concept .Even if it is line it puts scrollbar horizontally.
@rnveach

**Comment by rnveach** [comment]
> if we want content to wrap we have to apply css on pre tag which is child tag of source .
if we want scrollbar horizontally we have to apply css on source tag which is parent tag.

That still doesn't mean they be can't be separated by a `class` attribute.

Example CSS:
````
type1.source pre { .... }
type2.source { .... }
````

Example HTML:
````
<source class="type1"></source>
<source class="type2"></source>
````

@romani Please confirm you are ok with this change as it is different than what we discussed.

**Comment by sagar-shah94** [comment]
@rnveach Can you please restart wrecker/build?
I dont know why it is failing because mvn clean verify showed BUILD SUCCESS in my local.

**Comment by rnveach** [comment]
We can't restart wercker. It is a known issue with them. You will have to amend and push a new commit.

**Comment by sagar-shah94** [comment]
Ok sir .thanks will amend message and push again.

**Comment by romani** [comment]
@sagar-shah94 , please rebase you branch on most latest our master code to resolve wercker problem.

**Comment by romani** [comment]
lets keep two concepts of source blocks: wrapped and scrolled.

github do the same:
` this is very long line ....................... single line ............ . ... ...... ... single line ... single line ... single line ... single line... single line ... single line ... single line `

```
here is 
   multi line  with loooooooooooooooooooong looooooooooooong looooooonngg looooooooooooooooooooog line
```

Real examples:
`mvn exec:java -Dexec.mainClass="com.puppycrawl.tools.checkstyle.Main -Dexec.args="-c /sun_checks.xml src/main/java "`

```
JAVADOC -> <acronym title="as sooooooooooooooooooooooooooooooooooon as possible">ASAP</acronym><EOF> [0:0]
|--HTML_ELEMENT -> <acronym title="as soon as possible">ASAP</acronym> [0:0]
|   `--HTML_TAG -> <acronym title="as soon as possible">ASAP</acronym> [0:0]
|       |--HTML_ELEMENT_OPEN -> <acronym title="as soon as possible"> [0:0]
|       |   |--OPEN -> < [0:0]
|       |   |--HTML_TAG_NAME -> acronym [0:1]
|       |   |--WS ->   [0:8]
|       |   |--ATTRIBUTE -> title="as soon as possible" [0:9]
|       |   |   |--HTML_TAG_NAME -> title [0:9]
|       |   |   |--EQUALS -> = [0:14]
|       |   |   `--ATTR_VALUE -> "as soon as possible" [0:15]
|       |   `--CLOSE -> > [0:37]
|       |--TEXT -> ASAP [0:38]
|       `--HTML_ELEMENT_CLOSE -> </acronym> [0:42]
|           |--OPEN -> < [0:42]
|           |--SLASH -> / [0:43]
|           |--HTML_TAG_NAME -> acronym [0:44]
|           `--CLOSE -> > [0:51]
`--EOF -> <EOF> [0:52]
```

**Comment by rnveach** [comment]
@sagar-shah94 ping
Since we are out of proposal deadline, do you think you can finish this?
Don't forget to rebase on latest master.

**Comment by sagar-shah94** [comment]
@rnveach yeah currently fixing this issue.

**Comment by sagar-shah94** [comment]
@rnveach @romani 
Created 2 separate css .for code scrollbar will be there and for content wraping will happen.
Please verify and approve.

**Comment by sagar-shah94** [comment]
@rnveach suggested changes done applied css on suggested line numbers.please verify and approve.

**Comment by rnveach** [comment]
@sagar-shah94 What about my other comment?

> Please review other files too. writingchecks.xml was not changed but it has things similar to writingjavadocchecks.xml.vm.

These 2 files are similar. Why are there no changes in one, but changes in the other?

**Comment by sagar-shah94** [comment]
@rnveach made changes in writingchecks.xml.
Sorry missed reading that line of comment that is why forgot to make changes before

File: src/xdocs/cmdline.xml.vm [hunk 2]
```diff
@@ -33,7 +33,7 @@
     <section name="Command line usage">
       <p>
 
-      <source>
+      <source class="wrap-content">
```
Comment (romani) [inline src/xdocs/cmdline.xml.vm]: please keep it scrollable for this case. we need multi-line here.

File: src/xdocs/cmdline.xml.vm [hunk 3]
```diff
@@ -162,24 +162,24 @@ java -cp checkstyle-${projectVersion}-all.jar com.puppycrawl.tools.checkstyle.gu
...
-        <source>
+        <source class="wrap-content">
 git clone https://github.com/checkstyle/checkstyle.git
 cd checkstyle
 mvn clean compile
         </source>
 
         Run validation with arguments:
-        <source>
+        <source class="wrap-content">
```
Comment (romani) [inline src/xdocs/cmdline.xml.vm]: please remove from source " \"s and combine in one line, ones we have wrapping we could keep it as one line.
Please do the same for all other cases below.

File: src/xdocs/writingchecks.xml [hunk 4]
```diff
@@ -581,14 +581,14 @@ public class MethodLimitCheck extends AbstractCheck
       </p>
 
       <p>For Linux/Unix OS:</p>
-      <source>
+      <source class="wrap-content">
 java -classpath mycompanychecks.jar:checkstyle-${projectVersion}-all.jar \
```
Comment (romani) [inline src/xdocs/writingchecks.xml]: please remove from source " "s and combine in one line, ones we have wrapping we could keep it as one line.
Please do the same for all other cases with " \" in this file.

**Comment by rnveach** [comment]
> please share your opinion on wrapping of xml content

@romani I think I prefer it scrolled. Wrapping makes it harder to find the start and end of the line for the tag.

**Comment by sagar-shah94** [comment]
@romani @rnveach 
please verify and let me know if i am missing anything.

File: src/xdocs/cmdline.xml.vm [hunk 5]
```diff
@@ -203,7 +201,7 @@ java -jar target/checkstyle-X.X-SNAPSHOT-all.jar -c /sun_checks.xml MyClass.java
           <code>docs/sun_checks.xml</code> on a filesystem
         </b>
       </p>
-      <source>
+      <source class="wrap-content">
```
Comment (romani) [inline src/xdocs/cmdline.xml.vm]: if you put wrap-content, then in content of tag should not be any " \". Fix all cases

Comment (sagar-shah94) [inline src/xdocs/cmdline.xml.vm]: you want me to combine all content wherever wrap content css is used?

Comment (romani) [inline src/xdocs/cmdline.xml.vm]: yes

Comment (sagar-shah94) [inline src/xdocs/cmdline.xml.vm]: @romani removed spaces from all tags on which css is applied.

**Comment by romani** [comment]
Symbol "\" has to removed completely as it is line wrap symbol in command line, as you make them single line it is not required anymore

**Comment by sagar-shah94** [comment]
@romani done. please verify

**Comment by sagar-shah94** [comment]
@rnveach your suggested changes are done, please verify and approve.

**Pull request merged**